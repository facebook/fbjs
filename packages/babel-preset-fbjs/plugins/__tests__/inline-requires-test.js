/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

'use strict';

const babel = require('@babel/core');
const inlineRequiresPlugin = require('../inline-requires');
const pluginTester = require('babel-plugin-tester');

pluginTester({
  plugin: inlineRequiresPlugin,
  pluginName: 'inline-requires',
  pluginOptions: {
    inlineableCalls: ['customStuff'],
  },
  tests: {
    'inlines single usage': {
      code: ['var foo = require("foo");', 'foo.bar()'].join('\n'),
      snapshot: true,
    },

    'inlines multiple usages': {
      code: ['var foo = require("foo");', 'foo.bar()', 'foo.baz()'].join('\n'),
      snapshot: true,
    },

    'inlines any number of variable declarations': {
      code: [
        'var foo = require("foo"), bar = require("bar"), baz = 4;',
        'foo.method()',
      ].join('\n'),
      snapshot: true,
    },

    'ignores requires that are not assigned': {
      code: ['require("foo");'].join('\n'),
      snapshot: false,
    },

    'delete unused requires': {
      code: ['var foo = require("foo");'].join('\n'),
      snapshot: true,
    },

    'ignores requires that are re-assigned': {
      code: ['var foo = require("foo");', 'foo = "bar";'].join('\n'),
      snapshot: false,
    },

    'inlines requires that are referenced before the require statement': {
      code: [
        'function foo() {',
        '  bar();',
        '}',
        'var bar = require("baz");',
        'foo();',
        'bar();',
      ].join('\n'),
      snapshot: true,
    },

    'inlines destructured require properties': {
      code: [
        'var tmp = require("./a");',
        'var a = tmp.a',
        'var D = {',
        '  b: function(c) { c ? a(c.toString()) : a("No c!"); },',
        '};',
      ].join('\n'),
      snapshot: true,
    },

    'inlines functions provided via `inlineableCalls`': {
      code: [
        'const inlinedCustom = customStuff("foo");',
        'const inlinedRequire = require("bar");',
        '',
        'inlinedCustom();',
        'inlinedRequire();',
      ].join('\n'),
      snapshot: true,
    },
  },
});

describe('inline-requires', () => {
  const transform = (source, options) =>
    babel.transform(source.join('\n'), {
      ast: true,
      compact: true,
      plugins: [
        [require('@babel/plugin-transform-modules-commonjs'), {strict: false}],
        [inlineRequiresPlugin, options],
      ],
    });

  const compare = (input, output, options) => {
    expect(transform(input, options).code).toBe(
      transform(output, options).code,
    );
  };

  it('should be compatible with other transforms like transform-modules-commonjs', function() {
    compare(
      ['import Imported from "foo";', 'console.log(Imported);'],
      [
        'var _foo = _interopRequireDefault(require("foo"));',
        'function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }',
        'console.log(_foo.default);',
      ],
    );
  });

  it('should be compatible with `transform-modules-commonjs` when using named imports', function() {
    compare(
      [
        'import {a} from "./a";',
        'var D = {',
        '  b: function(c) { c ? a(c.toString()) : a("No c!"); },',
        '};',
      ],
      [
        'var D = {',
        '  b: function (c) {',
        '    c ? (0, require("./a").a)(c.toString()) : (0, require("./a").a)("No c!");',
        '  }',
        '};',
      ],
    );
  });

  it('should remove loc information from nodes', function() {
    const ast = transform(['var x = require("x"); x']).ast;
    const expression = ast.program.body[0].expression;

    function expectNoLocation(node) {
      expect(node.start).toBeUndefined();
      expect(node.end).toBeUndefined();
      expect(node.loc).toBeUndefined();
    }

    expectNoLocation(expression);
    expectNoLocation(expression.arguments[0]);
  });
});
