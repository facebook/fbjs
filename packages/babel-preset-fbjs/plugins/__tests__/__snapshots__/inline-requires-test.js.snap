// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inline-requires delete unused requires: delete unused requires 1`] = `
"
var foo = require(\\"foo\\");

      ↓ ↓ ↓ ↓ ↓ ↓


"
`;

exports[`inline-requires inlines any number of variable declarations: inlines any number of variable declarations 1`] = `
"
var foo = require(\\"foo\\"), bar = require(\\"bar\\"), baz = 4;
foo.method()

      ↓ ↓ ↓ ↓ ↓ ↓

var baz = 4;
require(\\"foo\\").method();
"
`;

exports[`inline-requires inlines destructured require properties: inlines destructured require properties 1`] = `
"
var tmp = require(\\"./a\\");
var a = tmp.a
var D = {
  b: function(c) { c ? a(c.toString()) : a(\\"No c!\\"); },
};

      ↓ ↓ ↓ ↓ ↓ ↓

var D = {
  b: function (c) {
    c ? require(\\"./a\\").a(c.toString()) : require(\\"./a\\").a(\\"No c!\\");
  }
};
"
`;

exports[`inline-requires inlines functions provided via \`inlineableCalls\`: inlines functions provided via \`inlineableCalls\` 1`] = `
"
const inlinedCustom = customStuff(\\"foo\\");
const inlinedRequire = require(\\"bar\\");

inlinedCustom();
inlinedRequire();

      ↓ ↓ ↓ ↓ ↓ ↓

customStuff(\\"foo\\")();
require(\\"bar\\")();
"
`;

exports[`inline-requires inlines multiple usages: inlines multiple usages 1`] = `
"
var foo = require(\\"foo\\");
foo.bar()
foo.baz()

      ↓ ↓ ↓ ↓ ↓ ↓

require(\\"foo\\").bar();
require(\\"foo\\").baz();
"
`;

exports[`inline-requires inlines requires that are referenced before the require statement: inlines requires that are referenced before the require statement 1`] = `
"
function foo() {
  bar();
}
var bar = require(\\"baz\\");
foo();
bar();

      ↓ ↓ ↓ ↓ ↓ ↓

function foo() {
  require(\\"baz\\")();
}

foo();
require(\\"baz\\")();
"
`;

exports[`inline-requires inlines single usage: inlines single usage 1`] = `
"
var foo = require(\\"foo\\");
foo.bar()

      ↓ ↓ ↓ ↓ ↓ ↓

require(\\"foo\\").bar();
"
`;
